### PR TITLE
LibJS: Implement RegExp legacy static properties

### PR DESCRIPTION
**Pull request by leeight**
https://github.com/tc39/proposal-regexp-legacy-features

### CODE REVIEW

File: Userland/Libraries/LibJS/Runtime/RegExpConstructor.cpp [hunk 1]
```diff
@@ -27,6 +27,44 @@ void RegExpConstructor::initialize(GlobalObject& global_object)
     define_native_accessor(*vm.well_known_symbol_species(), symbol_species_getter, {}, Attribute::Configurable);
 
     define_direct_property(vm.names.length, Value(2), Attribute::Configurable);
+
+    initialize_static_props(global_object);
+}
+
+void RegExpConstructor::initialize_static_props([[maybe_unused]] GlobalObject& global_object)
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Runtime/RegExpConstructor.cpp:34]: - please avoid the abbreviation, we don't have a function name length limit
- please link to the spec here
- no need to name the parameter if it's not used, or no need for maybe_unused if it is.

File: Userland/Libraries/LibJS/Runtime/RegExpConstructor.cpp [hunk 2]
```diff
@@ -93,4 +131,69 @@ JS_DEFINE_NATIVE_FUNCTION(RegExpConstructor::symbol_species_getter)
...
 }
 
+JS_DEFINE_NATIVE_FUNCTION(RegExpConstructor::input_setter)
+{
+    auto new_value = vm.argument(0);
+    global_object.regexp_constructor()->m_input = new_value;
+    return new_value;
+}
+
+// RegExp.input / RegExp.$_ [[Get]]
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Runtime/RegExpConstructor.cpp:141]: Please add links to these as well, just the name is hardly useful.

File: Userland/Libraries/LibJS/Runtime/RegExpConstructor.cpp [hunk 3]
```diff
@@ -93,4 +131,69 @@ JS_DEFINE_NATIVE_FUNCTION(RegExpConstructor::symbol_species_getter)
...
+    return global_object.regexp_constructor()->m_left_context;
+}
+
+// RegExp.rightContext / RegExp.$' [[Get]]
+JS_DEFINE_NATIVE_FUNCTION(RegExpConstructor::right_context_getter)
+{
+    return global_object.regexp_constructor()->m_right_context;
+}
+
+ThrowCompletionOr<Value> RegExpConstructor::get_captured_value(VM& vm, [[maybe_unused]] GlobalObject& global_object, size_t index)
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Runtime/RegExpConstructor.cpp:171]: get_foo() is usually not compliant with Serenity's code style, instead, it should be just foo()

File: Userland/Libraries/LibJS/Runtime/RegExpConstructor.cpp [hunk 4]
```diff
@@ -93,4 +131,69 @@ JS_DEFINE_NATIVE_FUNCTION(RegExpConstructor::symbol_species_getter)
...
+{
+    auto array = TRY(m_captured_values);
+    if (!array) {
+        return js_string(vm, "");
+    }
+    return array->get(index);
+}
+
+#define GEN_CAPTURED_VALUE_GETTER(index)                             \
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Runtime/RegExpConstructor.cpp:180]:
```suggestion
#define GENERATE_CAPTURED_VALUE_GETTER(index)                        \
```

File: Userland/Libraries/LibJS/Runtime/RegExpConstructor.h [hunk 5]
```diff
@@ -17,13 +18,42 @@ class RegExpConstructor final : public NativeFunction {
...
+    virtual void initialize_static_props(GlobalObject&);
     virtual bool has_constructor() const override { return true; }
-
+    ThrowCompletionOr<Value> get_captured_value(VM&, GlobalObject&, size_t);
+    ThrowCompletionOr<Array*> m_captured_values { nullptr };
+    ThrowCompletionOr<Value> m_input { js_string(vm(), "") };
+    ThrowCompletionOr<Value> m_last_match { js_string(vm(), "") };
+    ThrowCompletionOr<Value> m_last_paren { js_string(vm(), "") };
+    ThrowCompletionOr<Value> m_right_context { js_string(vm(), "") };
+    ThrowCompletionOr<Value> m_left_context { js_string(vm(), "") };
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Runtime/RegExpConstructor.h:40]: No one is visiting these, so they'll be deleted if a GC collect event happens, also, I'm not sure why the these are stored as ThrowCompletionOr, the exception shouldn't be stored here.

File: Userland/Libraries/LibJS/Runtime/RegExpPrototype.cpp [hunk 6]
```diff
@@ -344,10 +344,25 @@ static ThrowCompletionOr<Value> regexp_builtin_exec(GlobalObject& global_object,
...
     }
 
+    auto left_context = string.substring_view(0, match_index);
+    auto right_context = string.substring_view(end_index);
+
     // 22. Perform ! CreateDataPropertyOrThrow(A, "input", S).
     // NOTE: This step is performed last to allow the string to be moved into the js_string invocation.
     MUST(array->create_data_property_or_throw(vm.names.input, js_string(vm, move(string))));
 
+    // Set the capturedValues to RegExpConstructor
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Runtime/RegExpPrototype.cpp:354]: Please link to the proposal here and add proper spec comments

File: Userland/Libraries/LibJS/Runtime/RegExpPrototype.h [hunk 7]
```diff
@@ -20,8 +20,8 @@ class RegExpPrototype final : public PrototypeObject<RegExpPrototype, RegExpObje
 
 public:
     explicit RegExpPrototype(GlobalObject&);
-    virtual void initialize(GlobalObject&) override;
-    virtual ~RegExpPrototype() override = default;
+    virtual void initialize(GlobalObject&);
+    virtual ~RegExpPrototype() = default;
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Runtime/RegExpPrototype.h:24]: Keep the override keyword.

File: Userland/Libraries/LibJS/Tests/builtins/RegExp/RegExp.prototype.exec.js [hunk 8]
```diff
@@ -16,6 +16,7 @@ test("basic unnamed captures", () => {
     expect(res.length).toBe(2);
     expect(res[0]).toBe("fooooo");
     expect(res[1]).toBe("ooooo");
+    expect(RegExp.$1).toBe("ooooo");
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Tests/builtins/RegExp/RegExp.prototype.exec.js:19]: I think it's probably better to test this in its own separate test file? not too sure

File: Userland/Libraries/LibJS/Tests/builtins/RegExp/RegExp.prototype.test.js [hunk 9]
```diff
@@ -92,6 +92,19 @@ test("override exec with non-function", () => {
     expect(re.test("test")).toBe(true);
 });
 
+test("basic unnamed captures", () => {
```
Comment (alimpfard) [inline Userland/Libraries/LibJS/Tests/builtins/RegExp/RegExp.prototype.test.js:95]: ditto

File: Userland/Libraries/LibJS/Runtime/RegExpConstructor.cpp (hunk 1 above)
Comment (linusg) [reply Userland/Libraries/LibJS/Runtime/RegExpConstructor.cpp:34]: There's no point in making this a separate function in the first place, can all happen in `initialize()`

File: Userland/Libraries/LibJS/Runtime/RegExpConstructor.cpp [hunk 10]
```diff
@@ -93,4 +131,69 @@ JS_DEFINE_NATIVE_FUNCTION(RegExpConstructor::symbol_species_getter)
     return vm.this_value(global_object);
 }
 
+JS_DEFINE_NATIVE_FUNCTION(RegExpConstructor::input_setter)
+{
+    auto new_value = vm.argument(0);
+    global_object.regexp_constructor()->m_input = new_value;
+    return new_value;
```
Comment (linusg) [inline Userland/Libraries/LibJS/Runtime/RegExpConstructor.cpp:138]: This still needs to follow the spec - the one on GitHub we linked, not some random Microsoft docs. In this case:

https://github.com/tc39/proposal-regexp-legacy-features#set-regexpinput--val

So you'll have to implement the `SetLegacyRegExpStaticProperty` AO, which includes observable steps like throwing depending on the this value and calling ToString on the value being set.

**Review by linusg — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibJS/Runtime/RegExpConstructor.h (hunk 5 above)
Comment (leeight) [reply Userland/Libraries/LibJS/Runtime/RegExpConstructor.h:38]: OK

**Comment by leeight** [comment]
> This is still far from being able to be merged. Please take a look around LibJS to familiarize yourself with the existing style:
> 
> * Functions map to spec AOs - for some reason you just appended some steps of `UpdateLegacyRegExpStaticProperties` to `RegExpBuiltinExec`, and `SetLegacyRegExpStaticProperty` to the `RegExp.input` setter?
> * Spec comments have numbers
> * Everything new needs to have spec comments, this includes all of the getter/setter functions
> * Once again, please link to the actual spec. MDN is not the spec, neither is Microsoft developer docs

Ok, thanks for the feedback.

**Pull request closed without merging**